### PR TITLE
Connected navigation tabs and added guard locks

### PR DESCRIPTION
**Pull request by amandaadam**
In this branch, the following was done:
- Connected navigation tabs to the correct level pages
- Resolved a bug that saved the textbox value from the previous level in the next level 
- Added guard mechanisms for unlocked levels 
- Make it so that the program remembers the highest level that was completed

### CODE REVIEW

File: client/src/components/Header.tsx [hunk 1]
```diff
@@ -30,16 +31,17 @@ const Header = () => {
...
-    <Tabs>
+    <Tabs index={game.level}>
       <Flex bg="gray.200" w="100%" position="sticky" zIndex="1" p={4}>
         <TabList>
           <Tab onClick={() => game.jumpToLevel(0)}>Level 1</Tab>
-          <Tab onClick={() => game.jumpToLevel(1)}>Level 2</Tab>
-          <Tab onClick={() => game.jumpToLevel(2)}>Level 3</Tab>
-          <Tab onClick={() => game.jumpToLevel(3)}>Level 4</Tab>
-          <Tab onClick={() => game.jumpToLevel(4)}>Level 5</Tab>
+          <Tab isDisabled={game.hasSeenLevel(0)} onClick={() => game.jumpToLevel(1)}>Level 2</Tab>
```
Comment (Aradhya-B) [inline client/src/components/Header.tsx:38]: Semantically, it doesn't make sense for the tab to be `isDisabled` when the user has indeed seen the previous level. I would change the function name to `hasNotSeenLevel`.

Comment (Aradhya-B) [reply client/src/components/Header.tsx:38]: Actually, using `Not` in a function name is usually indicative that your logic is flipped (you'll have a double negative when you need the `true` case). Keep the name as `hasSeenLevel` but refactor the function to be semantically correct and then make this `isDisabled={!game.hasSeenLevel(0)}`.

File: client/src/context/GameContext.tsx [hunk 2]
```diff
@@ -93,6 +96,7 @@ export const GameProvider: React.FC = ({ children }) => {
   const [stepIndex, setStepIndex] = useState(0);
   const [numElems, setNumElems] = useState(10);
   const [[min, max], setMinMax] = useState([0, 20]);
+  const [maxLevelSeen, setMaxLevelSeen] = useState<number>(0);
```
Comment (Aradhya-B) [inline client/src/context/GameContext.tsx:99]: nit: typescript is pretty good at interpreting types, don't need the `<number>` here. also just nicer for the sake of consistency with the states defined in the lines above.

File: client/src/context/GameContext.tsx [hunk 3]
```diff
@@ -190,13 +194,32 @@ export const GameProvider: React.FC = ({ children }) => {
...
+    setCorrect({});
+  }
+
+  function hasSeenLevel(lastLevel: number) {
+      if (lastLevel < maxLevelSeen) {
+          return false;
+      }
+      return true;
+    }
+
```
Comment (Aradhya-B) [inline client/src/context/GameContext.tsx:210]: So really, you have "seen" the `lastLevel` when `lastLevel < maxLevelSeen`, not the other way around as is defined right now. So really it should return `true` when `lastLevel < maxLevelSeen`, however, this whole method may be simplified to `return lastLevel < maxLevelSeen`.

File: client/src/pages/Home.tsx [hunk 4]
```diff
@@ -1,6 +1,7 @@
 import React, { useState } from "react";
 import { Button, Select, VStack, Center, Text } from "@chakra-ui/react";
 import { useNavigate } from "react-router-dom";
+import { useGame } from "@/context/GameContext";
```
Comment (Aradhya-B) [inline client/src/pages/Home.tsx:4]: nit: remove unused import

File: client/src/pages/Home.tsx [hunk 5]
```diff
@@ -11,8 +12,11 @@ const Home: React.FC<HomeProps> = ({}) => {
 
   const [selection, setSelection] = useState<number>(0);
 
+  const game = useGame();
```
Comment (Aradhya-B) [inline client/src/pages/Home.tsx:15]: nit: we're not using game anymore, so we can remove it

File: client/src/components/Header.tsx [hunk 6]
```diff
@@ -22,6 +22,7 @@ export type Props = {};
 const Header = () => {
   // TODO: Remove this hook once next level is properly implemented.
   // Tabs should not be clickable, this is only for demo purpose.
+  //index={game.level}
```
Comment (Aradhya-B) [inline client/src/components/Header.tsx:25]: nit: remove this comment

**Review by Aradhya-B — CHANGES_REQUESTED** [review]
Great work guys! Just requested a few code changes, everything looks good functionally.

File: client/src/components/Header.tsx [hunk 7]
```diff
@@ -30,16 +30,17 @@ const Header = () => {
...
         <TabList>
           <Tab onClick={() => game.jumpToLevel(0)}>Level 1</Tab>
-          <Tab onClick={() => game.jumpToLevel(1)}>Level 2</Tab>
-          <Tab onClick={() => game.jumpToLevel(2)}>Level 3</Tab>
-          <Tab onClick={() => game.jumpToLevel(3)}>Level 4</Tab>
-          <Tab onClick={() => game.jumpToLevel(4)}>Level 5</Tab>
+          <Tab isDisabled={!game.hasSeenLevel(0)} onClick={() => game.jumpToLevel(1)}>Level 2</Tab>
+          <Tab isDisabled={!game.hasSeenLevel(1)} onClick={() => game.jumpToLevel(2)}>Level 3</Tab>
+          <Tab isDisabled={!game.hasSeenLevel(2)} onClick={() => game.jumpToLevel(3)}>Level 4</Tab>
+          <Tab isDisabled={!game.hasSeenLevel(3)} onClick={() => game.jumpToLevel(4)}>Level 5</Tab>
```
Comment (garethlau) [inline client/src/components/Header.tsx:40]: Be careful about this. `game.hasSeenLevel()` will only be called when the component mounts and may not reflect the current state of the variables. If this is producing the intended behaviour then go ahead.

**Comment by Aradhya-B** [comment]
looks good to me, thanks for making those changes. :)

**Review by Aradhya-B — APPROVED** [review]

File: client/src/context/GameContext.tsx [hunk 8]
```diff
@@ -47,12 +47,15 @@ interface ContextType {
   currStep: Step;
   stepIndex: number;
   level: number;
+  maxLevelSeen: number;
```
Comment (garethlau) [inline client/src/context/GameContext.tsx:50]: Does `maxLevelSeen` have to be exposed? Are there components that require this value if `hasSeenLevel` is provided?

**Review by garethlau — APPROVED** [review]
Great work!

**Pull request merged**